### PR TITLE
Blog app: add devise

### PR DESCRIPTION
**Pull request by najibullahjafari**
Here:
-  We removed the current_user method in ApplicationController, devise will provide us with one.

- A user can register a new user.
- A user logs in with a combination of email and password.
- A hashed password should be stored in the database.
- Ask for confirmation of email.
- A user can reset the password.

### CODE REVIEW

File: app/views/layouts/_navbar.html.erb [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+#   <%# if defined?(user_signed_in?) && user_signed_in? %>
+#     <%#= link_to "Sign Out", destroy_user_session_path, method: :delete %>
+#   <%# else %>
+#     <%#= link_to "Sign In", new_user_session_path %>
+#     <%#= link_to "Register", new_user_registration_path %>
+#   <%# end %>
+# </nav>
+<nav class="navbar navbar-expand-lg bg-light d-flex justify-content-end">
+  <%= button_to "Sign out", destroy_user_session_path, method: :delete, class: 'btn btn-primary' %>
 </nav>
```
Comment (cancelei) [inline app/views/layouts/_navbar.html.erb]: Here we have 2 nav tags, and some uncommented code.

I think we can refactor that to make it more readable/understandable.

**Review by cancelei — COMMENTED** [review]
Good job on the code @najibullahjafari & @Grandi0z !

I have 1 suggested change below.

**Review by cancelei — COMMENTED** [review]
OK.

File: app/views/layouts/_navbar.html.erb (hunk 1 above)
Comment (najibullahjafari) [reply app/views/layouts/_navbar.html.erb]: solved

File: app/models/user.rb [hunk 2]
```diff
@@ -1,13 +1,18 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
   # Associations
+  devise :database_authenticatable, :registerable, :confirmable, :recoverable, :rememberable, :validatable
```
Comment (mckent05) [inline app/models/user.rb]: - You have done well so far, but you seem to have configured 3 devise modules in your app, this necessarily does not affect the function of your application, but to ensure your code is clean and easier to read and maintain, please configure just one devise module. The config on line 7 seems to have all the required modules for this milestone. Please fix this 🙏 by removing the config on line 4 & 15 👍 👍

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @najibullahjafari, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- You have done well adding devise to your application 👍 👍 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ghost — APPROVED** [review]
Hi Team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

 



## Highlights

✅ Devise is setup correctly 👍
✅ A user can log in with a combination of email and password.
 

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
   has_many :posts, foreign_key: 'author_id'
   has_many :comments
   has_many :likes
 
   # Validations
   validates :name, presence: true
   validates :posts_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
-
+  devise :confirmable
   # Methods
```
Comment (omar-muhamad) [inline app/models/user.rb]: - [x] You already added the `confirmable` module in line 2 there is no need to repeat it again in line 10. please remove line 10 to achieve the `DRY` principle as the previous reviewer suggested.

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @cancelei,

#### Nice job making the required changes. Unfortunately, you missed some of them but you are almost there.



## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [x] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Comment by omar-muhamad** [comment]
Hi @cancelei and @frank1738,

 [Importamt notice] It seems that you and your team members submitted the same PR, This is a peer-programming milestone but every team member should submit his own project as per the [work submission mode](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_devise.md#work-and-submission-mode).

 I continued with my review because you (Glauber Bannwart) are the GitHub account owner and the review was submitted with your name.
 The approved review above seemed to be submitted by one of your team members, So please don't repeat this again.

**Comment by cancelei** [comment]
Cancelei is GLauber Bannwart, but I am having trouble to approve this task on my dashboard.

**Review by Qoosim — APPROVED** [review]
#  *Status: Approved* :tada: :heavy_check_mark: 

Hi @najibullahjafari, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
-  Implemented required changes. :heavy_check_mark: 
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by nicupop729** [comment]
Hi team, 
My name is Nicolae Pop, and I am a Senior Code Reviewer at Microverse.

Please note that for this milestone, the requirement was to work in pair-programming on the repositories of all team members and submit the PR for review individually, using your own repository.

Please always check the [Work and Submission Mode](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/blog_app_devise.md#work-and-submission-mode) section in the project requirements.

image


@najibullahjafari, the approval from your dashboard for this milestone will be reverted. Please work as a team to complete this milestone in your repository too, and when ready, please ask again for a review using a PR from your own repository.

Good Luck!:four_leaf_clover:

**Pull request merged**